File: DXHtmlEditor/DXHtmlEditorWebView.cs
using System;
using System.Threading.Tasks;
using DevExpress.Utils;
using Microsoft.Web.WebView2.Core;
using Microsoft.Web.WebView2.WinForms;

namespace DXHtmlEditor {
    public class DXHtmlEditorWebView : WebView2, IDXHtmlEditorAPI {
        DXHtmlEditorClient client;
        public DXHtmlEditorWebView() {
            if(!DesignMode)
                Initialize();
        }
        #region API
        public async Task<string> GetHtmlText() {
            await EnsureIsLoaded();
            return await client.GetHtmlText();
        }
        public async Task<string> SetHtmlText(string htmlString) {
            await EnsureIsLoaded();
            return await client.SetHtmlText(htmlString);
        }
        public bool CanSetTheme(string themeName) {
            return client.ThemeName != themeName;
        }
        public async Task SetTheme(string themeName) {
            string htmlText = await GetHtmlText();
            StartLoading();
            client.ThemeName = themeName;
            await SetHtmlText(htmlText);
            Refresh();
        }
        #endregion API
        #region Custom API Sample
        public async Task Undo() {
            await EnsureIsLoaded();
            await client.Undo();
        }
        public async Task Redo() {
            await EnsureIsLoaded();
            await client.Redo();
        }
        #endregion
        #region Events
        static readonly object htmlChanged = new object();
        static readonly object htmlLoaded = new object();
        //
        public event EventHandler HtmlChanged {
            add { Events.AddHandler(htmlChanged, value); }
            remove { Events.RemoveHandler(htmlChanged, value); }
        }
        public event EventHandler HtmlLoaded {
            add { Events.AddHandler(htmlLoaded, value); }
            remove { Events.RemoveHandler(htmlLoaded, value); }
        }
        void IDXHtmlEditorAPI.RaiseOnHtmlChanged() {
            (Events[htmlChanged] as EventHandler)?.Invoke(this, EventArgs.Empty);
        }
        void IDXHtmlEditorAPI.RaiseOnLoaded() {
            loadedSource?.TrySetResult(true);
            (Events[htmlLoaded] as EventHandler)?.Invoke(this, EventArgs.Empty);
        }
        #endregion Events
        #region Internal
        void Initialize() {
            StartLoading();
            client = new DXHtmlEditorClient(this);
            CoreWebView2InitializationCompleted += OnCoreWebView2InitializationCompleted;
        }
        protected override void Dispose(bool disposing) {
            CoreWebView2InitializationCompleted -= OnCoreWebView2InitializationCompleted;
            loadedSource?.TrySetCanceled();
            loadedSource = null;
            DisposeHelper.Dispose(ref client);
            base.Dispose(disposing);
        }
        protected override void OnHandleCreated(EventArgs e) {
            base.OnHandleCreated(e);
            if(!DesignMode)
                EnsureCoreWebView2Async();
        }
        protected override void OnHandleDestroyed(EventArgs e) {
            base.OnHandleDestroyed(e);
            loadedSource?.TrySetCanceled();
        }
        void OnCoreWebView2InitializationCompleted(object sender, CoreWebView2InitializationCompletedEventArgs e) {
            client.Initialize(CoreWebView2);
        }
        TaskCompletionSource<bool> loadedSource;
        void StartLoading() {
            var prevSource = loadedSource;
            loadedSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
            if(prevSource != null && !prevSource.Task.IsCompleted)
                prevSource.TrySetCanceled();
        }
        async Task EnsureIsLoaded() {
            try {
                Task<bool> loadingTask = loadedSource?.Task;
                bool loaded = await loadingTask;
                if(!loaded || loadingTask.IsFaulted)
                    throw new Exception("DXHtmlEditor loading failed", loadingTask?.Exception);
            }
            catch(TaskCanceledException e) {
                if(e.Task != loadedSource?.Task)
                    await EnsureIsLoaded();
            }
        }
        #endregion
    }
}

File: VB/DXHtmlEditor/DXHtmlEditorWebView.vb
Imports System
Imports System.Threading.Tasks
Imports DevExpress.Utils
Imports Microsoft.Web.WebView2.Core
Imports Microsoft.Web.WebView2.WinForms

Namespace DXHtmlEditor

    Public Class DXHtmlEditorWebView
        Inherits WebView2
        Implements IDXHtmlEditorAPI

        Private client As DXHtmlEditorClient

        Public Sub New()
            If Not DesignMode Then Initialize()
        End Sub

#Region "API"
        Public Async Function GetHtmlText() As Task(Of String)
            Await EnsureIsLoaded()
            Return Await client.GetHtmlText()
        End Function

        Public Async Function SetHtmlText(ByVal htmlString As String) As Task(Of String)
            Await EnsureIsLoaded()
            Return Await client.SetHtmlText(htmlString)
        End Function

        Public Function CanSetTheme(ByVal themeName As String) As Boolean
            Return Not Equals(client.ThemeName, themeName)
        End Function

        Public Async Function SetTheme(ByVal themeName As String) As Task
            Dim htmlText As String = Await GetHtmlText()
            StartLoading()
            client.ThemeName = themeName
            Await SetHtmlText(htmlText)
            Refresh()
        End Function

#End Region  ' API
#Region "Custom API Sample"
        Public Async Function Undo() As Task
            Await EnsureIsLoaded()
            Await client.Undo()
        End Function

        Public Async Function Redo() As Task
            Await EnsureIsLoaded()
            Await client.Redo()
        End Function

#End Region
#Region "Events"
        Private Shared ReadOnly htmlChangedField As Object = New Object()

        Private Shared ReadOnly htmlLoadedField As Object = New Object()

        '
        Public Custom Event HtmlChanged As EventHandler
            AddHandler(ByVal value As EventHandler)
                Events.AddHandler(htmlChangedField, value)
            End AddHandler

            RemoveHandler(ByVal value As EventHandler)
                Events.RemoveHandler(htmlChangedField, value)
            End RemoveHandler

            RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
            End RaiseEvent
        End Event

        Public Custom Event HtmlLoaded As EventHandler
            AddHandler(ByVal value As EventHandler)
                Events.AddHandler(htmlLoadedField, value)
            End AddHandler

            RemoveHandler(ByVal value As EventHandler)
                Events.RemoveHandler(htmlLoadedField, value)
            End RemoveHandler

            RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
            End RaiseEvent
        End Event

        Private Sub RaiseOnHtmlChanged() Implements IDXHtmlEditorAPI.RaiseOnHtmlChanged
            TryCast(Events(htmlChangedField), EventHandler)?.Invoke(Me, EventArgs.Empty)
        End Sub

        Private Sub RaiseOnLoaded() Implements IDXHtmlEditorAPI.RaiseOnLoaded
            loadedSource?.TrySetResult(True)
            TryCast(Events(htmlLoadedField), EventHandler)?.Invoke(Me, EventArgs.Empty)
        End Sub

#End Region  ' Events
#Region "Internal"
        Private Sub Initialize()
            StartLoading()
            client = New DXHtmlEditorClient(Me)
            AddHandler CoreWebView2InitializationCompleted, AddressOf OnCoreWebView2InitializationCompleted
        End Sub

        Protected Overrides Sub Dispose(ByVal disposing As Boolean)
            RemoveHandler CoreWebView2InitializationCompleted, AddressOf OnCoreWebView2InitializationCompleted
            loadedSource?.TrySetCanceled()
            loadedSource = Nothing
            DisposeHelper.Dispose(client)
            MyBase.Dispose(disposing)
        End Sub

        Protected Overrides Sub OnHandleCreated(ByVal e As EventArgs)
            MyBase.OnHandleCreated(e)
            If Not DesignMode Then EnsureCoreWebView2Async()
        End Sub

        Protected Overrides Sub OnHandleDestroyed(ByVal e As EventArgs)
            MyBase.OnHandleDestroyed(e)
            loadedSource?.TrySetCanceled()
        End Sub

        Private Sub OnCoreWebView2InitializationCompleted(ByVal sender As Object, ByVal e As CoreWebView2InitializationCompletedEventArgs)
            client.Initialize(CoreWebView2)
        End Sub

        Private loadedSource As TaskCompletionSource(Of Boolean)

        Private Sub StartLoading()
            Dim prevSource = loadedSource
            loadedSource = New TaskCompletionSource(Of Boolean)(TaskCreationOptions.RunContinuationsAsynchronously)
            If prevSource IsNot Nothing AndAlso Not prevSource.Task.IsCompleted Then prevSource.TrySetCanceled()
        End Sub

        Private Async Function EnsureIsLoaded() As Task
            Try
                Dim loadingTask As Task(Of Boolean) = loadedSource?.Task
                Dim loaded As Boolean = Await loadingTask
                If Not loaded OrElse loadingTask.IsFaulted Then Throw New Exception("DXHtmlEditor loading failed", loadingTask?.Exception)
            Catch e As TaskCanceledException
                If e.Task IsNot loadedSource?.Task Then Await EnsureIsLoaded()
            End Try
        End Function
#End Region
    End Class
End Namespace
